util-taglib/src/com/liferay/taglib/util/InlineUtil.java

```diff
@@ -33,18 +35,40 @@ public static String buildDynamicAttributes(
 
 		StringBundler sb = new StringBundler(dynamicAttributes.size() * 4);
 
+		List<String> attributesList = Arrays.asList(_HTML_BOOLEAN_ATTRIBUTES);
+
 		for (Map.Entry<String, Object> entry : dynamicAttributes.entrySet()) {
 			String key = entry.getKey();
 
 			if (!key.equals("class")) {
-				sb.append(key);
-				sb.append("=\"");
-				sb.append(String.valueOf(entry.getValue()));
-				sb.append("\" ");
+				if (attributesList.contains(key)) {
+					boolean value = (boolean)entry.getValue();
+
+					if (value) {
+						sb.append(key);
+						sb.append("=\"");
+						sb.append(key);
```

Comment:
Why would we set the value of `boolean` attributes to their key? We should either take `value` or nothing. Taking `value` would be the most conservative approach since it keeps whatever the user intended to pass (even if we know it to be unnecessary).

Based on this, I think this could be equivalent to (untested):

```java
if (attributesList.contains(key)) {
    boolean value = (boolean)entry.getValue();

    if (!value) {
        return;
    }

    // Logic to append attributes is the same
    sb.append(key);
    ...
}
```

Reply:
> Why would we set the value of boolean attributes to their key? 

From what I understand, setting a boolean attribute to true is invalid, but you're right that we should keep whatever the user intended to pass, so I'll try that now.

Reply:
> [...] setting a boolean attribute to true is invalid [...]

Can you back this up with the place in the spec that specifies it?

I'm pretty sure that claim is true, but browser implementors don't complain about this which makes me prefer to keep values passed by a user.

```html
<button disabled="something">...</button>
```

```javascript
console.log($0.disabled); // true
console.log($0.getAttribute('disabled')); // 'something'

$0.disabled = 'something_else'; // 'something_else'
console.log($0.disabled); // true
console.log($0.getAttribute('disabled'); // ''

$0.setAttribute('disabled', 'something_else'); // 'something_else'
console.log($0.disabled); // true
console.log($0.getAttribute('disabled'); // 'something_else'
```

The `disabled` getter/setter seems to normalize values, but the semantics of `getAttribute/setAttribute` are maintained, so I'd still side with letting users pass values to boolean attributes to avoid breaking any possible queries/selectors that might already be in use.

/cc @wincent, do you have any thoughts on this?

Reply:
> Can you back this up with the place in the spec that specifies it?

Yes of course, it's [here](https://html.spec.whatwg.org/#boolean-attribute),
![](https://user-images.githubusercontent.com/5572/107934611-94ed2a80-6f80-11eb-8bfe-851cf0cbd80a.png)

But I agree with you on just setting the value to whatever was passed if it's not "false"

Reply:
> do you have any thoughts on this?

Well, the "safest" change is probably to preserve the value that the user passed (safest in the sense that the behavior should be basically the same as it was before in terms of what makes it to the browser).

The more "correct" change would evidently be to slavishly adhere to the spec, which @julien has linked to. FWIW:

- If you do `$(thing).attr('disabled', true)`, it will set `disabled="disabled"` in the DOM.
- If you do `thing.disabled = true` or `thing.setAttribute('disabled', '')`, it will show up in the inspector as an attribute without a value, but `thing.getAttribute('disabled')` shows that the attribute's value is indeed `''` in both cases.

I personally find the empty string behavior a bit too subtle (especially the whole boolean/string coercion thing), so I like the jQuery way better, and I'd vote for that.